src/gtkspellcheck/spellcheck.py

```diff
@@ -149,32 +151,121 @@ def exists(self, language):
             return language in self.mapping
 
     class _Mark:
-        def __init__(self, buffer, name, start):
+        def __init__(self, buffer, name, start, iter_worker):
             self._buffer = buffer
             self._name = name
             self._mark = self._buffer.create_mark(self._name, start, True)
+            self._iter_worker = iter_worker
 
         @property
         def iter(self):
             return self._buffer.get_iter_at_mark(self._mark)
 
         @property
         def inside_word(self):
-            return self.iter.inside_word()
+            return self._iter_worker.inside_word(self.iter)
 
         @property
         def word(self):
             start = self.iter
-            if not start.starts_word():
-                start.backward_word_start()
+            if not self._iter_worker.starts_word(start):
+                self._iter_worker.backward_word_start(start)
             end = self.iter
-            if end.inside_word():
-                end.forward_word_end()
+            if self._iter_worker.inside_word(end):
+                self._iter_worker.forward_word_start(end)
             return start, end
 
         def move(self, location):
             self._buffer.move_mark(self._mark, location)
 
+    class _IterWorker:
+        def __init__(self, extra_word_chars):
+            self._extra_word_chars = extra_word_chars
+
+        def is_extra_word_char(self, loc):
+            # Language extra chararacters should also be processed once Enchant's
+            # enchant_dict_get_extra_word_characters is exposed in PyEnchant
+
+            char = loc.get_char()
+            return char != "" and char in self._extra_word_chars
+
+        def inside_word(self, loc):
+            if loc.inside_word():
+                return True
+            elif self.starts_word(loc):
+                return True
+            elif loc.ends_word() and not self.ends_word(loc):
+                return True
+            else:
+                return False
+
+        def starts_word(self, loc):
+            if loc.starts_word():
+                if loc.is_start():
+                    return True
+                else:
+                    tmp = loc.copy()
+                    tmp.backward_char()
+                    return not self.is_extra_word_char(tmp)
+            else:
+                return False
+
+        def ends_word(self, loc):
+            if loc.ends_word():
+                if loc.is_end():
+                    return True
+                else:
+                    tmp = loc.copy()
+                    tmp.forward_char()
+                    return not self.is_extra_word_char(tmp)
+            else:
+                return False
+
+        def forward_word_end(self, loc):
+            def move_through_extra_chars():
+                moved = False
+                while self.is_extra_word_char(loc):
+                    moved = True
```

Comment:
Can we put this after the `if` statement?

Otherwise, this function may return `true` despite not moving the location.

Reply:
Sure, great, changed :+1: